packages/core/src/storages/request_provider.ts

```diff
@@ -395,13 +408,17 @@ export abstract class RequestProvider implements IStorage {
             return resultsToReturn;
         };
 
-        const initialChunk = sources.splice(0, batchSize);
-
         // Add initial batch of `batchSize` to process them right away
+        const initialChunk = await chunksIterator.peek();
+        if (initialChunk === undefined) {
+            return { addedRequests: [], waitForAllRequestsToBeAdded: Promise.resolve([]) };
+        }
+
         const addedRequests = await attemptToAddToQueueAndAddAnyUnprocessed(initialChunk);
+        await chunksIterator.next();
 
-        // If we have no more requests to add, return early
-        if (!sources.length) {
+        // If we have no more requests to add, return immediately
```

Comment:
Generators return `{ done: true }` when there's nothing more to yield. afaiac, we're only using `peek` here to see if we're at the end, sooo... do we really need it?

Reply:
We don't _really need_ it, but it's a nice abstraction :shrug: Here, we just want to bail out early if there is nothing left in the iteratorable, and if we just used `.next()`, we'd need to handle both cases - the one where it's empty and the one where we get a value and need to handle that, _plus_ whatever is left in the iterable. It's not hard per se to do it, but it's kinda... gritty.